asteroid/masknn/attention.py

```diff
@@ -0,0 +1,229 @@
+import torch.nn.functional as F
+import torch.nn as nn
+from torch.nn.modules.activation import MultiheadAttention
+from asteroid.masknn import activations, norms
+import torch
+from asteroid.utils import has_arg
+
+class ImprovedTransformedLayer(nn.Module):
+    def __init__(self, d_model, n_heads, dim_ff, dropout=0., activation="relu", bidirectional=True, norm="gLN"):
+        super(ImprovedTransformedLayer, self).__init__()
+
+        self.mha = MultiheadAttention(d_model, n_heads, dropout=dropout)
+        self.recurrent = nn.LSTM(d_model, dim_ff, bidirectional=bidirectional)
+        self.dropout = nn.Dropout(dropout)
+        ff_inner_dim = 2*dim_ff if bidirectional else dim_ff
+        self.linear = nn.Linear(ff_inner_dim, d_model)
+        self.activation = activations.get(activation)()
+        self.norm_mha = norms.get(norm)(d_model)
+        self.norm_ff = norms.get(norm)(d_model)
+
+    def forward(self, x):
+        x = x.transpose(1, -1)
+        # x is batch, seq_len, channels
+        # self-attention is applied
+        out = self.mha(x, x, x)[0]
+        x = self.dropout(out) + x
+        x = self.norm_mha(x.transpose(1, -1)).transpose(1, -1)
+
+        # lstm is applied
+        out = self.linear(self.dropout(self.activation(self.recurrent(x)[0])))
+        x = self.dropout(out) + x
+        return self.norm_ff(x.transpose(1, -1))
+
+
+class Oladd(nn.Module):
+
+    def __init__(self, chunk_size, hop_size):
+        super(Oladd, self).__init__()
+        self.chunk_size = chunk_size
+        self.hop_size = hop_size
+        self.n_orig_frames = None
+
+    def unfold(self, x):
+        # x is (batch, chan, frames)
+        batch, chan, frames = x.size()
+        assert x.ndim == 3
+        self.n_orig_frames = x.shape[-1]
+        unfolded = torch.nn.functional.unfold(
+            x.unsqueeze(-1),
+            kernel_size=(self.chunk_size, 1),
+            padding=(self.chunk_size, 0),
+            stride=(self.hop_size, 1),
+        )
+
+        return unfolded.reshape(batch, chan, self.chunk_size, -1) # (batch, chan, chunk_size, n_chunks)
+
+    def fold(self, x):
+        # x is (batch, chan, chunk_size, n_chunks)
+        batch, chan, chunk_size, n_chunks = x.size()
+        to_unfold = x.reshape(batch, chan * self.chunk_size, n_chunks)
+        x = torch.nn.functional.fold(
+            to_unfold,
+            (self.n_orig_frames, 1),
+            kernel_size=(self.chunk_size, 1),
+            padding=(self.chunk_size, 0),
+            stride=(self.hop_size, 1),
+        )
+
+        x /= (self.chunk_size / self.hop_size)
```

Comment:
This is to unit test against `fold(unfold(x)) == x` right?

Reply:
yes this guarantees perfect reconstruction